### PR TITLE
Addition of U-Net Model

### PR DESCRIPTION
**Pull request by aaronrockmenezes**
## Description

Adding the U-Net segmentation model

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by shreyasvinaya** [comment]
@aaronrockmenezes seems like the torch tests are failing:
```bash
FAILED deepchem/models/torch_models/tests/test_unet.py::test_UNetModel - RuntimeError: Given input size: (512x1x1). Calculated output size: (512x0x0). Output size is too small
FAILED deepchem/models/torch_models/tests/test_unet.py::test_restore_UNetModel - RuntimeError: Given input size: (512x1x1). Calculated output size: (512x0x0). Output size is too small
```

File: deepchem/models/torch_models/unet.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        x = self.decoder2(torch.cat([x, x2], 1))
+        x = self.upsample(x)
+        x = self.decoder1(torch.cat([x, x1], 1))
+
+        # Output
+        x = self.output(x)
+        return x
+
+
+class UNetModel(TorchModel):
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/unet.py:102]: please add docs and example to this

File: deepchem/models/torch_models/unet.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+from deepchem.models.torch_models.modular import TorchModel
+from deepchem.models.losses import BinaryCrossEntropy
+import torch
+import torch.nn as nn
+
+
+class UNet(nn.Module):
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/unet.py:7]: please add docs and example to this

File: deepchem/models/torch_models/unet.py (hunk 1 above)
Comment (aaronrockmenezes) [reply deepchem/models/torch_models/unet.py:104]: Done

File: deepchem/models/torch_models/unet.py (hunk 2 above)
Comment (aaronrockmenezes) [reply deepchem/models/torch_models/unet.py:8]: Done

Comment (rbharath) [reply deepchem/models/torch_models/unet.py:8]: Also add this to the docs/ as a new model. And add it to the table of models

File: deepchem/models/torch_models/tests/test_unet.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    # Using NumpyDataset for testing
+    np_dataset = dc.data.NumpyDataset(input_samples, output_samples)
+
+    unet_model = UNetModel(in_channels=3, out_channels=1)
+
+    unet_model.fit(np_dataset, nb_epoch=1)
+    pred = unet_model.predict(np_dataset)
+
+    assert pred.shape == output_samples.shape
+
```
Comment (rbharath) [inline deepchem/models/torch_models/tests/test_unet.py:25]: Please add an overfit test

Comment (aaronrockmenezes) [reply deepchem/models/torch_models/tests/test_unet.py:25]: Added

**Comment by aaronrockmenezes** [comment]
Need to add model to model cheatsheet

File: deepchem/models/torch_models/unet.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+        # Upsampling
+        self.upsample = nn.Upsample(scale_factor=2,
+                                    mode='bilinear',
+                                    align_corners=True)
+
+        # Output
+        self.output = nn.Conv2d(64, out_channels, kernel_size=1)
+
+    def conv_block(self, in_channels, out_channels):
```
Comment (rbharath) [inline deepchem/models/torch_models/unet.py:89]: Please add type annotations

File: deepchem/models/torch_models/unet.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        out_channels: int
+            Number of output channels.
+        """
+        return nn.Sequential(
+            nn.Conv2d(in_channels, out_channels, kernel_size=3, padding='same'),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(out_channels, out_channels, kernel_size=3,
+                      padding='same'), nn.ReLU(inplace=True))
+
+    def forward(self, x):
```
Comment (rbharath) [inline deepchem/models/torch_models/unet.py:104]: Same here (types)

File: docs/source/api_reference/models.rst [hunk 6]
```diff
@@ -555,6 +555,12 @@ TextCNNModel
 .. autoclass:: deepchem.models.torch_models.TextCNNModel
   :members:
 
+UNetModel
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:558]: Can you add a screenshot to show the docs render?

**Review by rbharath — COMMENTED** [review]
A few minor issues; ready to merge after those addressed

**Comment by aaronrockmenezes** [comment]
image
image

**Comment by aaronrockmenezes** [comment]
image
This is the UNet model as part of the general model cheatsheet

File: deepchem/models/torch_models/unet.py (hunk 5 above)
Comment (aaronrockmenezes) [reply deepchem/models/torch_models/unet.py]: Done

File: docs/source/api_reference/models.rst (hunk 6 above)
Comment (aaronrockmenezes) [reply docs/source/api_reference/models.rst:558]: Done

**Pull request merged**